---
title: "Introduction to the scFeatureFilter package"
author: "Vignette Authors"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to scFeatureFilter}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Preamble
## Package aim
One early step of single cell RNA-seq analysis is the filtering of features (i.e. genes or transcripts)
to retain only the most expressed ones. This allow to focus on features less affected by technical
variations (relatively), and often increase the power of subsequent analyses.

This step is usually done by applying one or a few arbitrary thresholds (i.e. mean TPM > 1),
or by using data from spike-in RNA.
This package aims at informing that threshold decision, especially in the cases
when spike-in RNA are not available

## Method summary
Highly expressed features are relatively less affected by technical variations that lowly expressed ones,
and thus better reflect biological variation
We propose to use a small bin of highly expressed features as a reference to estimate the relative decrease 
of biological variation in bins of features of decreasing expression. Indeed, features from major
transcription programs will be correlated with each other across bins.

One of the package first step is to select this reference bin of features (or 'top window' of features).
We propose to use the 100 features with the highest mean expression across cell and a reasonably low
fraction of 0 expression as reference. Using more features will dilute the amount of biological variation
reflected in this reference bin (as well as increase computation time), while using less features might 
result in insufficient capture of the biological complexity of the samples.
Subsequent bins of features of decreasing mean expression across cells are then defined (by default by 
slice of 1000 features). The size of the subsequent feature bins matters much less than the size of the
reference bin.
Several negative control bins are defined by shuffling values of the top bin expression matrix.
Then every feature in each bin is correlated to every feature in the reference bin and the control bins
to obtain correlation coefficient distributions.
The variation of the distributions of correlation coefficients can be used to inform the threshold decision.

## Installation
Instructions to arrive shortly.

Once installed, the package is ready to be loaded:
```{r}
library(scFeatureFilter)

# we then edit the plot theme
library(ggplot2)
theme_set(theme_bw())
```


# Integrated usage
You can use the package in a single function call, that will filter the expression matrix
to keep only the most informative highly expressed features:
```{r}
# example dataset included with the package:
scData_hESC

# filtering of the dataset with a single function call:
sc_feature_filter(scData_hESC)
```


# Detailed usage
## Loading data
*scFeatureFilter* need an expression matrix either as a `matrix`, a `data.frame`, a `tibble` or a
`SingleCellExperiment` R object.
Features should be in rows, while cells should be in columns. 

We recommend to provide expression values that were normalised to correct from
library size, for example TPM or FPKM, rather than raw counts.

Support of Bioconductor's expression set objects will arrive shortly.
An example dataset is supplied with the package:
```{r}
scData_hESC
```
*The example dataset was processed by [Mantsoki et al. (2016)](https://www.ncbi.nlm.nih.gov/pubmed/26951854),  data from human Embryonic Stem cells was generated by [Yan et al. (2013)](https://www.ncbi.nlm.nih.gov/pubmed/23934149).*

## Binning data
Firstly, we add a mean expression and a coefficient of variation column to the expression matrix
using the `calculate_cvs` function:
```{r}
calculate_cvs(scData_hESC)
```

We can explore the mean - variance relationship of the dataset with the `plot_mean_variance` function:
```{r collapse=TRUE}
library(magrittr) # to use the pipe %>%

calculate_cvs(scData_hESC) %>%
    plot_mean_variance(colourByBin = FALSE)
```

We then define the top window (or reference bin) with the `define_top_genes` function, and the other bins withe
the `bin_scdata` function:
```{r collapse=TRUE}
scData_hESC %>%
    calculate_cvs %>%
    define_top_genes(window_size = 100) %>%
    bin_scdata(window_size = 1000)
```

We can plot the resulting bin definition using the same `plot_mean_variance` function:
```{r collapse=TRUE}
myPlot <- scData_hESC %>%
    calculate_cvs %>%
    define_top_genes(window_size = 100) %>%
    bin_scdata(window_size = 1000) %>%
    plot_mean_variance(colourByBin = TRUE, density_color = "blue")

myPlot
```

Note that the outputs of this package plotting functions are `ggplot2` objects that can be 
further customized:
```{r}
myPlot + annotation_logticks(sides = "l")
```

## Correlation of the bins
We can now correlate every feature in each window to every other feature in the reference bin and the
randomized control bins, using the function `correlate_windows`:
```{r collapse=TRUE}
corDistrib <- scData_hESC %>%
    calculate_cvs %>%
    define_top_genes(window_size = 100) %>%
    bin_scdata(window_size = 1000) %>%
    correlate_windows(n_random = 3)
```

Distributions of the correlation values can be visualized using the `correlations_to_densities` and `plot_correlations_distributions` functions:
```{r collapse=TRUE}
corDens <- correlations_to_densities(corDistrib, absolute_cc = TRUE)
plot_correlations_distributions(corDens, facet_ncol = 5)
```

The coloured lines indicate the correlation coefficient distributions for each bins compared to 
the reference bin (bin 1 show the autocorrelation of the reference bin), while the thicker blue 
lines with grey area (barely visible) represents the correlation coefficient distributions for
each bins compared to the randomized control bins.

Notice how the correlation distributions shift from high to low value as the bin number is increasing 
(i.e. as the  mean expression level of the features is decreasing).

*Note: the `absolute_cc = TRUE` option, which is the default, indicate that the package will work
with the absolute value of correlation coefficient. Therefore, highly negatively correlated features
will count as highly correlated. It also produces clearer plots, and reduce the emphasis on non-symmetrical,
near 0, shifts of correlation values that are not interpretable.*

We can quantify the steady decrease of correlation values as feature expression goes down using the
`get_mean_median` function:
```{r}
metrics <- get_mean_median(corDistrib)
metrics
plot_correlations_distributions(corDens, metrics = metrics, facet_ncol = 5)
```

The added dashed lines are the means of the distributions. The `plot_metric` function focalized on these means of distributions:
```{r}
plot_metric(metrics, show_ctrl = FALSE, show_threshold = FALSE)
```

The bar represents the mean value of the correlation coefficient distribution of each bin 
against the reference bin, while the dot and whiskers are against the randomized control windows.
The mean of the correlation distributions is decreasing steadily, and reach a plateau after bin 11.

## Choosing a threshold
On the previous plot, we can add a background reference dashed line by taking the mean value of the 
correlations against the randomized control bins for every bin:
```{r}
plot_metric(metrics, show_ctrl = TRUE, show_threshold = FALSE)
```

And we can decide to keep only bins for which the mean correlation value against the reference window 
is higher that, for example, twice the background:
```{r}
plot_metric(metrics, show_ctrl = TRUE, show_threshold = TRUE, threshold = 2)
```

The function `determine_bin_cutoff` will return the number of their first bin of feature that
fell below our threshold:
```{r}
determine_bin_cutoff(metrics, threshold = 2)
```

The function `filter_expression_table` can be used to obtain a filtered expression matrix:
```{r}
bined_data <- scData_hESC %>%
    calculate_cvs %>%
    define_top_genes(window_size = 100) %>%
    bin_scdata(window_size = 1000)
metrics <- correlate_windows(bined_data, n_random = 3) %>%
    get_mean_median

filtered_data <- filter_expression_table(
    bined_data,
    bin_cutoff = determine_bin_cutoff(metrics)
)

dim(scData_hESC)
dim(filtered_data)
filtered_data

```

*Note: most of the features are filtered at the `calculate_cvs` step because they have a too high
fraction of 0 expression (maximum fraction of 0 expression is 75% by default).*

## Working with SingleCellExperiment objects
This package accpets `SingleCellExperiment` as inputs for `sc_feature_filter`
and `calculate_cvs` functions. However it does not retun such objects at the moment.
One can retrieve a filtered `SingleCellExperiment` objects like this:
```{r}
library(SingleCellExperiment)
library(scRNAseq) # example datasets
data("allen")
sce_allen <- as(allen, "SingleCellExperiment")

# sce_allen is an SingleCellExperiment object
sce_allen

filtered_allen <- sc_feature_filter(sce_allen, sce_assay = "rsem_tpm")
is.matrix(filtered_allen) # filtered_allen is a tibble

sce_filtered_allen <- sce_allen[rownames(filtered_allen), ]
sce_filtered_allen
```



